lgc/patch/PatchBufferOp.cpp

```diff
@@ -1475,6 +1475,71 @@ static unsigned getLSIntrinsicID(bool isLoad, bool isInvariant, unsigned accessS
   return Intrinsic::amdgcn_raw_buffer_store;
 }
 
+// =====================================================================================================================
+// Recalculate offset by moving the constant part of the offset from "Value" component to "unsigned" component.
+//
+// Analyze the GEP contributing to "Value" component, and move the constant part to the "unsigned" component.
+//
+// @param initialOffset : The initial offset, consisting of (Value, unsigned) components.
+// @returns : The updated offset, where some of "Value" component has been moved to "unsigned" component.
+std::pair<Value *, unsigned> PatchBufferOp::recalculateOffset(std::pair<Value *, unsigned> initialOffset) {
+
+  auto ptrToInt = dyn_cast<PtrToIntInst>(initialOffset.first);
+  if (!ptrToInt)
+    return initialOffset;
+
+  auto bitcast = dyn_cast<BitCastInst>(ptrToInt->getPointerOperand());
+  Value *gepCand = bitcast ? bitcast->getOperand(0) : ptrToInt->getPointerOperand();
+
+  auto GEP = dyn_cast<GetElementPtrInst>(gepCand);
+  if (!GEP)
+    return initialOffset;
+
+  // Find the splitting point in GEP: first const index in the series of all const indices.
+  auto idxEnd = GEP->idx_end();
+  auto idxFirstConstInSequence = idxEnd;
+
+  for (auto idxIt = GEP->idx_begin(); idxIt != idxEnd; ++idxIt) {
+    if (!isa<ConstantInt>(*idxIt)) {
+      idxFirstConstInSequence = idxIt + 1;
+    }
+  }
+
+  if (idxFirstConstInSequence == idxEnd)
+    return initialOffset;
```

Comment:
I think what you're trying to do here is best described as "find the (possibly empty) tail of constant indices and split it off". But in order to do that correctly, the for loop needs to go in reverse.